### ISSUE TITLE
Not working with macOS High Sierra

### ISSUE THREAD

**Issue by mteros** [MichaelDimmitt/Aerial_Desktop#34]
After upgrading to macOS High Sierra, program is not working anymore.  Installation seems to go without a hitch but nothing happens after startup or waking from sleep.

**Comment by MichaelDimmitt** [comment]
@mteros, yep you are correct. I first ran into this a month ago #33 
unfortunately this is a Macintosh computer bug.

With High Sierra, operating system has changed dramatically.
This application uses the mac default  Screensaver Engine 

The location of the engine has moved. 

the engine still works but the --background flag has completely messed up!
I have verified this by moving the engine from a working old operating system and running that application on the new operating system. I know the issue is with the operating system and not the program because the result is the same. 

It still runs as a background process but does not run in the background. (see below test)
reported this a month ago to apple, they marked it as a duplicate and have not otherwise responded.

The below test does not cause any damage to your computer.
the answer to the test in the below comment will determine if mac has fixed it. 
warning, if the computer gets stuck, apple did it... but simply close the laptop or sleep shutdown the computer and it will end the background process. <b>This does not cause any damage to your computer.</b>
this is a test and not in the application.  

Until they fix the issue:
I do not want the launch agent to wake up the program after your sleep ends its execution.

**Comment by MichaelDimmitt** [comment]
## simply close the laptop or sleep shutdown the computer if stuck
<b>warning this is the test mentioned above
<br>it does not cause any damage to your computer:</b>

putting this into your terminal runs this mac application.
`/System/Library/CoreServices/./ScreenSaverEngine.app/Contents/MacOS/ScreenSaverEngine -background`

if it still works message me! and I can fix everything. If it causes trouble put your computer to sleep your computer will be fine. and for resolving the issue sadly lets wait a month or two. 
if you run into trouble my cell is 904 200 6567 please call for a valid reason.

**Comment by MichaelDimmitt** [comment]
@mteros, I will update immediately when I discover the issue is fixed.
Hoping to construct a workaround to solve the problem. But my few attempts have failed. 

It may occur that apple fixes the problem. Which I will report and hope for seeing that my issue was marked a duplicate. 

Lately my focus has been turning this application into a menubar and adding the coolest screensavers. For instance, I have one that plays youtube videos!

So I hope this inconvenience is not a deal breaker and apple or myself will resolve the issue.

**Comment by quantumgolem** [comment]
@MichaelDimmitt Hey, I haven't been following along with the updated. Does your fork have the ability to set custom videos as the screen saver? :)

**Comment by MichaelDimmitt** [comment]
This really would be best in a new issue since it is a different subject matter.
@sn0wyfall, not current behavior to hook up your own videos but in the future you could hook up your YouTube playlist. Which could have your own custom videos of infinite length!

Also I could look at aerial's code. It has a website it downloads the videos from so if you put your video files online it could access it off of the website you point it to.

Personally I think the YouTube solution is more user friendly to hook up. And in a menubar it will be easy to toggle between YouTube or aerial with a click to operate back and forth.

I have to rewrite some swift code to make the YouTube user experience better. Which is why I haven't released it. But I currently have that working in a different repo on a public GitHub organization.

**Comment by mteros** [comment]
Hi Michael,

Thanks for the update.  Let’s hope Apple will come to their senses.

Minh

**Comment by quantumgolem** [comment]
@MichaelDimmitt All right, thanks a lot! I think you can already use your own videos with Aerial by changing the video files on your drive to the ones you want to use. Haven't tried it yet though!

**Comment by chrisdobler** [comment]
What if Apple never adds that support? I could see them declaring it uses too much battery or just not caring about it. I wonder what the difficulty level would be to patching the screensaver engine.

**Comment by MichaelDimmitt** [comment]
Good News this githubber named @Waitsnake . Figured this one out.
credit: https://github.com/Waitsnake/AnimatedGif/commit/1a2369c45ef104cff22a496065e36ef379dad7bb and https://github.com/Waitsnake/AnimatedGif/issues/28
The change will be coming soon and I will have everyone up and running. @plicjo

**Comment by MichaelDimmitt** [comment]
the new `-window` command does the trick. 
this command in the terminal on high sierra... 
works and does not cause the computer to be locked out:
```bash 
/System/Library/CoreServices/ScreenSaverEngine.app/Contents/MacOS/ScreenSaverEngine -window -background 
```
the only challenge left for this issue is how to resize the window to full screen.
which i am investigating. ive seen the objective c solution. with waitsnake
but his program uses objective c from the inside 
and my application uses applescript and bash from the outside to open the background window.

It seems to be a terminal window. it says terminal preferences when that window is selected.
`/System/Library/CoreServices/ScreenSaverEngine.app/Contents/MacOS/ScreenSaverEngine -window`

All of this information is from my work on MacOS High Sierra 
10.13.3 ps, I heard a rumor that it is fixed without a workaround in 10.13.4

so keep your ears peeled for the apple update.

**Comment by MichaelDimmitt** [comment]
going to use 
```bash
if[ 10.13.4 < $(sw_vers -productVersion) ]
```
stick to the old way that doesnt work on high sierra but works with older operating systems.
```bash
/System/Library/Frameworks/ScreenSaver.framework/Resources/./ScreenSaverEngine.app/Contents/MacOS/ScreenSaverEngine -background

else  /System/Library/CoreServices/ScreenSaverEngine.app/Contents/MacOS/ScreenSaverEngine -window -background 
fi
```
to wrap the command to get product version into a test
to ensure it is always working.

**Comment by plicjo** [comment]
WOOHOO! This app is awesome. I can't wait to see it popping again.

**Comment by Waitsnake** [comment]
Hope the rumors are true and Apple fix this problem with the "-background" parameter. 

Even if AnimatedGif uses the parameter "-window" as workaround there is an other issue to be aware of. When using the parameter "-window" and the option "show with clock" at the same time the screensaver engine crashes, see [AnimatedGif/issues/30](https://github.com/Waitsnake/AnimatedGif/issues/30). As workaround therefore I disable the option "show with clock" before starting the screensaver engine with "-window" option. But this is no ideal solution anyway.

Aerial screensaver is a cool project and videos it shows are stunning.

**Comment by MichaelDimmitt** [comment]
@Waitsnake, see https://github.com/JohnCoates/Aerial

Soon im going to post a menubar project I made that does youtube video screensavers and many more screensavers. Id be happy to include yours if you are interested. Or show more about the menubar project.

email: michaelgdimmitt@gmail.com

**Comment by MichaelDimmitt** [comment]
Update on this, I have AerialDesktop working in a degraded fashion. 
That enhancement can be tracked here: [github.com/MichaelDimmitt/Aerial_Desktop/issues/38]
(https://github.com/MichaelDimmitt/Aerial_Desktop/issues/38)
Will close this issue when the new pull request gets implemented.
[github.com/MichaelDimmitt/Aerial_Desktop/pull/37](https://github.com/MichaelDimmitt/Aerial_Desktop/pull/37)

And open a new pull request about fixing the degraded implementation.
